mongomock/aggregate.py

```diff
@@ -498,15 +498,86 @@ def _handle_lookup_stage(in_collection, database, options):
     foreign_field = options['foreignField']
     local_name = options['as']
     foreign_collection = database.get_collection(foreign_name)
-    for doc in in_collection:
+    out_doc = copy.deepcopy(in_collection)
+    for doc in out_doc:
         query = doc.get(local_field)
         if isinstance(query, list):
             query = {'$in': query}
         matches = foreign_collection.find({foreign_field: query})
-        # TODO(pascal): Do not modify the input collection.
         doc[local_name] = [foreign_doc for foreign_doc in matches]
 
-    return in_collection
+    return out_doc
+
+
+def _handle_graph_lookup_stage(in_collection, database, options):
+    if not isinstance(options.get('maxDepth', 0), six.integer_types):
+        raise OperationFailure(
+            "Argument 'maxDepth' to $graphLookup must be a number")
+    if not isinstance(options.get('restrictSearchWithMatch', {}), dict):
+        raise OperationFailure(
+            "Argument 'restrictSearchWithMatch' to $graphLookup must be a Dictionary")
+    if not isinstance(options.get('depthField', ''), six.string_types):
+        raise OperationFailure(
+            "Argument 'depthField' to $graphlookup must be a string")
+    if 'startWith' not in options:
+        raise OperationFailure(
+            "Must specify 'startWith' field for a $graphLookup")
+    for operator in ('as', 'connectFromField', 'connectToField', 'from'):
+        if operator not in options:
+            raise OperationFailure(
+                "Must specify '%s' field for a $graphLookup" % operator)
+        if not isinstance(options[operator], six.string_types):
+            raise OperationFailure(
+                "Argument '%s' to $graphLookup must be string" % operator)
+        if options[operator].startswith('$'):
+            raise OperationFailure("FieldPath field names may not start with '$'")
+        if operator in ('connectFromField', 'as') and \
+                '.' in options[operator]:
+            raise NotImplementedError(
+                "Although '.' is valid in the '%s' "
+                'parameter for the $graphLookup stage of the aggregation '
+                'pipeline, it is currently not implemented in Mongomock.' % operator)
+
+    foreign_name = options['from']
+    start_with = options['startWith']
+    connect_from_field = options['connectFromField']
+    connect_to_field = options['connectToField']
+    local_name = options['as']
+    max_depth = options.get('maxDepth', None)
+    depth_field = options.get('depthField', None)
+    restrict_search_with_match = options.get('restrictSearchWithMatch', {})
+    foreign_collection = database.get_collection(foreign_name)
+    out_doc = copy.deepcopy(in_collection)
```

Comment:
As I mentioned above, we might be able to avid a deepcopy: we need to copy two levels only, the list of doc and the first level. As this PR has been opened for a while, I suggest you keep a TODO for me to speed up the copy.